### PR TITLE
adding a patch item operation with a context and other patch / patch derived operation changes

### PR DESCRIPTION
**Pull request by shawkins**
## Description
This is to address #3067 the current patch operation with only the modified item is dangerous when there are concurrent modifications.

This requires a two argument patch call.

I looked at adding the resourceVersion to a GET call, but the kubernetes support is "no older than" - https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter so you are not guaranteed to get back the object of that version.  Based upon that, this is the simpler and backwards compatible change.

I'm not sure about the the appropriate way to clone the item in the patch call - or even if that is needed.  I can't see any guidance on whether it's expected that the operation may modify the item passed into it.

I also removed the retry logic on patching - it seems you will get a 422 error if the patch can't be applied rather than a conflict.

@rohanKanojia let me know if this is on the right track.

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by metacosm** [comment]
The way we clone in the Java operator SDK: https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java#L64-L71 (serialize -> deserialize)

**Comment by shawkins** [comment]
@metacosm thanks, I've added that clone method as a utility method on the Serialization class.

@rohanKanojia I see there are a lot of internal calls to the single argument path function - given that these have concurrency issues, is that something that needs addressed as well?

**Comment by shawkins** [comment]
@iocanel I see a relatively recent change to use edit(visitors) via a patch call.  I've updated this pr to base that upon a protected method to create the builder when possible - is there currently any way to find/construct the builder for a given type programatically so that this method is not needed?

@iocanel and others - what is the expectation for cloning via constructing a builder with a reference object?  for example - new PodBuilder(pod) - from what I can see in the code this is a shallow copy.  So for example in calls to edit where an item is associated with the context, that item may be edited.  Is this expected?  I see in some places this is referred to as a clone: https://github.com/fabric8io/kubernetes-client/blob/74cc63df9b6333d083ee24a6ff3455eaad0a6da8/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/DeploymentRollingUpdater.java#L41

@rohanKanojia another concern is that some of the calls to patch have overriding logic, such as to toggle the cascading flag - that logic will be applied in the case of a call to patch with an object, but it will not be applied when calling the new methods with a patch string.  Should there be common enforcement of that logic?

**Comment by shawkins** [comment]
> @iocanel and others - what is the expectation for cloning via constructing a builder with a reference object? for example - new PodBuilder(pod) - from what I can see in the code this is a shallow copy. 

Actually this seems to only be an issue for Resource derived from CustomResource (I just assumed it was the same for other classes like Pod). I'll spawn this off as a separate issue - #3076

**Comment by shawkins** [comment]
The other approach instead of introducing the two argument method would be to expose the withItem method on the Resource interface, so that you could make the patch call:

client.configMaps().inNamespace(currentNamespace).withName(name).withItem(base).patch(modified);

**Comment by rohanKanojia** [comment]
I don't remember why this `cascading` is used in combination with `patch`. Since cascading is deprecated, I don't think we should use it in our newer methods

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java [hunk 1]
```diff
@@ -48,12 +51,12 @@
...
   @BeforeClass
   public static void init() {
-    ClusterEntity.apply(CronJobIT.class.getResourceAsStream("/patch-it.yml"));
+    ClusterEntity.apply(PatchIT.class.getResourceAsStream("/patch-it.yml"));
   }
 
   @Before
   public void initNamespace() {
-    this.currentNamespace = session.getNamespace();
+    this.currentNamespace = ClusterEntity.getArquillianNamespace();
```
Comment (rohanKanojia) [inline kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java:59]: Why is this change required?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java [hunk 2]
```diff
@@ -312,4 +312,19 @@ private static String readSpecFileFromInputStream(InputStream inputStream) {
     }
     return JSON_MAPPER.readerFor(KubernetesResource.class).readValue(jsonString);
   }
+  
+  /**
+   * Create a copy of the resource via serialization
+   * @param <T>
+   * @param resource
+   * @return
+   */
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:321]: Please complete javadoc

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java (hunk 1 above)
Comment (shawkins) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java:59]: It may not be.  When I was running the test through my ide I was seeing setup and initNamespace using different namespaces - this made sure they were using the same one.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java (hunk 2 above)
Comment (shawkins) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:320]: It's updated now.

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java (hunk 1 above)
Comment (rohanKanojia) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java:59]: I remember I had added `ClusterEntity.getArquillianNamespace()` to find out active namespace used in tests from static context(where session is not available). I think `session.getNamespace()` should work for your use case.

Comment (shawkins) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java:59]: Perhaps because of killing the tests prematurely in the ide or something else, they can return different namespaces - so the setup happens in one namespace, while the session is in another.  Then referencing the session namespace here causes the test to fail.  I'll remove this change.

**Comment by rohanKanojia** [comment]
Some tests are failing. Are they related to this PR?

```
 Error:  Failures: 
Error:    SecurityContextConstraintsTest.testEdit:117 expected: <1> but was: <0>
Error:  Errors: 
Error:    OpenshiftRoleBindingTest.testPatchWithOnlySubjects:160 » NullPointer
Error:    OpenshiftRoleBindingTest.testPatchWithUserNamesAndGroupsAndNoSubjects:183 » NullPointer
Error:    OpenshiftRoleBindingTest.testPatchWithUserNamesAndGroupsAndOverwriteSubjects:204 » NullPointer
[INFO] 
Error:  Tests run: 744, Failures: 1, Errors: 3, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
```

**Comment by shawkins** [comment]
> Some tests are failing. Are they related to this PR?

Yes they are.  There's an assumption down in OperationSupport.checkName that a HasMetadata object will always return a non-null metadata.  However these and another test were not written this way.  I've now relaxed that assumption, rather than continuing to refine the tests.

**Review by rohanKanojia — APPROVED** [review]

**Review by metacosm — COMMENTED** [review]
LGTM apart from minor wording in exception message.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java [hunk 3]
```diff
@@ -42,18 +49,63 @@ public HasMetadataOperation(OperationContext ctx) {
...
+   * Get the current item from the server, consulting the metadata for the name if needed
+   */
+  protected Gettable<T> fromServer(ObjectMeta metadata) {
+    if (getName() != null) {
+      return fromServer();
+    }
+    if (metadata != null) {
+      return withName(metadata.getName()).fromServer();
+    }
+    throw new KubernetesClientException("Name not specified. But operation requires name.");
```
[the remark this answers is not in the document]
Comment (shawkins) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:92]: I took this wording from OperationSupport - do you want all of the similar messages updated?

[the remark this answers is not in the document]
Comment (metacosm) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:92]: Nevermind then. I found the wording strange but if it's already used elsewhere, let's keep it as-is.

**Review by metacosm — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java [hunk 4]
```diff
@@ -42,18 +49,63 @@ public HasMetadataOperation(OperationContext ctx) {
...
+  public T edit(Visitor... visitors) {
+    T item = getMandatory();
+    T clone = Serialization.clone(item);
+    return patch(null, clone, createVisitableBuilder(item).accept(visitors).build(), false);
+  }
+  
+  /**
+   * Get the current item from the server, consulting the metadata for the name if needed
+   */
+  protected Gettable<T> fromServer(ObjectMeta metadata) {
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:85]: why protected and not private?

Comment (shawkins) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:85]: It can be whatever you want.  BaseOperation.fromServer is public, but I was unsure if this should be as well, so I went with protected.

Comment (manusa) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:85]: I really don't see how a children class would need to reimplement this, anyway, let's leave it and get this merged :rocket:
